File: src/Sorting/HeapSort/heap_sort.py
def heap_sort(array):
    build_max_heap(array)
    heap_size = len(array) - 1
    for i in range(len(array) - 1, 1, -1):
        array[0], array[i] = array[i], array[0]
        heap_size = heap_size - 1
        max_heapify(array, i, heap_size)


def build_max_heap(array):
    for i in range(len(array) // 2, 0, -1):
        max_heapify(array, i, len(array) - 1)


def max_heapify(array, i, heap_size):
    left = get_left_position(i)
    right = get_right_position(i)
    largest = i

    if left < heap_size and array[left] > array[i]:
        largest = left

    if right < heap_size and array[right] > array[largest]:
        largest = right

    if largest != i:
        array[i], array[largest] = array[largest], array[i]
        max_heapify(array, largest, heap_size)


def get_left_position(i):
    return i * 2


def get_right_position(i):
    return i * 2 + 1

File: src/Sorting/HeapSort/HeapSort.ts
export function heapSort(array: number[]): void {
  buildMaxHeap(array);

  for (let i = array.length - 1; i >= 0; i--) {
    [array[0], array[i]] = [array[i], array[0]];
    maxHeapify(array, 0, i);
  }
}

export function buildMaxHeap(array: number[]): void {
  for (let i = Math.floor(array.length / 2) - 1; i >= 0; i--) {
    maxHeapify(array, i, array.length);
  }
}

function maxHeapify(array: number[], i: number, heapSize: number): void {
  while (i !== heapSize) {
    const left = getLeftPosition(i);
    const right = getRightPosition(i);
    let largest = i;
    if (left < heapSize && array[left] > array[largest]) {
      largest = left;
    }
    if (right < heapSize && array[right] > array[largest]) {
      largest = right;
    }

    if (largest === i) {
      return;
    }

    [array[i], array[largest]] = [array[largest], array[i]];
    maxHeapify(array, largest, heapSize);
  }
}

function getLeftPosition(i: number): number {
  return i * 2 + 1;
}

function getRightPosition(i: number): number {
  return i * 2 + 2;
}
